File: code/client/munkilib/munkistatus.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
munkistatus.py

Created by Greg Neagle on 2009-09-24.

Utility functions for using MunkiStatus.app
to display status and progress.
"""
from __future__ import absolute_import, print_function

import os
import time

# PyLint cannot properly find names inside Cocoa libraries, so issues bogus
# No name 'Foo' in module 'Bar' warnings. Disable them.
# pylint: disable=E0611
from Foundation import NSDistributedNotificationCenter
from Foundation import NSNotificationDeliverImmediately
from Foundation import NSNotificationPostToAllSessions
# pylint: enable=E0611

# we use lots of camelCase-style names. Deal with it.
# pylint: disable=C0103

# our NSDistributedNotification identifier
NOTIFICATION_ID = 'com.googlecode.munki.managedsoftwareupdate.statusUpdate'

# keep our current status. We keep this so that notification clients
# that come "online" late can get current state
_currentStatus = {}

def initStatusDict():
    '''Initialize our status dictionary'''
    global _currentStatus
    _currentStatus = {
        'message': '',
        'detail': '',
        'percent': -1,
        'stop_button_visible': True,
        'stop_button_enabled': True,
        'command': '',
        'pid': os.getpid()
    }

def launchMunkiStatus():
    '''Uses launchd KeepAlive path so it launches from a launchd agent
    in the correct context.
    This is more complicated to set up, but makes Apple (and launchservices)
    happier.
    There needs to be a launch agent that is triggered when the launchfile
    is created; and that launch agent then runs MunkiStatus.app.'''
    initStatusDict()

    launchfile = "/var/run/com.googlecode.munki.MunkiStatus"
    try:
        open(launchfile, 'w').close()
    except (OSError, IOError):
        pass
    time.sleep(0.1)
    if os.path.exists(launchfile):
        os.unlink(launchfile)


def postStatusNotification():
    '''Post a status notification'''
    dnc = NSDistributedNotificationCenter.defaultCenter()
    dnc.postNotificationName_object_userInfo_options_(
        NOTIFICATION_ID,
        None,
        _currentStatus,
        NSNotificationDeliverImmediately + NSNotificationPostToAllSessions)


def message(messageText):
    '''Sets the status message.'''
    _currentStatus['message'] = messageText
    postStatusNotification()


def detail(detailsText):
    '''Sets the detail text.'''
    _currentStatus['detail'] = detailsText
    postStatusNotification()


def percent(percentage):
    '''Sets the progress indicator to 0-100 percent done.
    If you pass a negative number, the progress indicator
    is shown as an indeterminate indicator (barber pole).'''
    _currentStatus['percent'] = percentage
    postStatusNotification()


def hideStopButton():
    '''Hides the stop button.'''
    _currentStatus['stop_button_visible'] = False
    postStatusNotification()


def showStopButton():
    '''Shows the stop button.'''
    _currentStatus['stop_button_visible'] = True
    postStatusNotification()


def disableStopButton():
    '''Disables (grays out) the stop button.'''
    _currentStatus['stop_button_enabled'] = False
    postStatusNotification()


def enableStopButton():
    '''Enables the stop button.'''
    _currentStatus['stop_button_enabled'] = True
    postStatusNotification()


def activate():
    '''Brings MunkiStatus window to the front.'''
    _currentStatus['command'] = 'activate'
    postStatusNotification()
    # now clear the command; unlike the other fields, this
    # should not persist between notifications
    _currentStatus['command'] = ''


def quit_app():
    '''Tells the status app that we're done.'''
    _currentStatus['command'] = 'quit'
    postStatusNotification()
    # now clear the command; unlike the other fields, this
    # should not persist between notifications
    _currentStatus['command'] = ''


def restartAlert():
    '''Tells MunkiStatus to display a restart alert.'''
    _currentStatus['command'] = 'showRestartAlert'
    postStatusNotification()
    # now clear the command; unlike the other fields, this
    # should not persist between notifications
    _currentStatus['command'] = ''


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/munkistatus.swift
//
//  munkistatus.swift
//  munki
//
//  Created by Greg Neagle on 8/6/24.
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

/// a Singleton class that contains our current status for display
class MunkiStatusInfo {
    static let shared = MunkiStatusInfo()

    var status: PlistDict

    private init() {
        status = [
            "message": "",
            "detail": "",
            "percent": -1,
            "stop_button_visible": true,
            "stop_button_enabled": true,
            "command": "",
            "pid": ProcessInfo.processInfo.processIdentifier,
        ]
    }
}

/// Uses launchd KeepAlive path so it launches from a launchd agent
/// in the correct context.
/// This is more complicated to set up, but makes Apple (and launchservices) happier.
/// There needs to be a launch agent that is triggered when the launchfile is created;
/// and that launch agent then runs MunkiStatus.app.
/// Note: this only works at the loginwindow, since the LaunchAgent is only
/// loaded at the loginwindow.
func munkiStatusLaunch() {
    let launchfile = "/var/run/com.googlecode.munki.MunkiStatus"
    if FileManager.default.createFile(atPath: launchfile, contents: nil) {
        usleep(1_000_000)
        try? FileManager.default.removeItem(atPath: launchfile)
    } else {
        printStderr("Couldn't create launchpath \(launchfile)")
    }
}

/// Post a status notification
func postStatusNotification() {
    let NOTIFICATION_NAME = "com.googlecode.munki.managedsoftwareupdate.statusUpdate"
    let dnc = DistributedNotificationCenter.default()
    dnc.postNotificationName(
        NSNotification.Name(rawValue: NOTIFICATION_NAME),
        object: nil,
        userInfo: MunkiStatusInfo.shared.status,
        options: [.deliverImmediately, .postToAllSessions]
    )
}

/// Sets the status message.
func munkiStatusMessage(_ text: String) {
    MunkiStatusInfo.shared.status["message"] = text
    postStatusNotification()
}

/// Sets the detail text.
func munkiStatusDetail(_ text: String) {
    MunkiStatusInfo.shared.status["detail"] = text
    postStatusNotification()
}

func munkiStatusPercent(_ percentage: Int) {
    MunkiStatusInfo.shared.status["percent"] = percentage
    postStatusNotification()
}

/// Sets the progress indicator to 0-100 percent done.
/// If you pass a negative number, the progress indicator
/// is shown as an indeterminate indicator.
func munkiStatusHideStopButton() {
    // Tells MunkiStatus/MSC to hide the stop button.
    MunkiStatusInfo.shared.status["stop_button_visible"] = false
    postStatusNotification()
}

/// Tells MunkiStatus/MSC to show the stop button.
func munkiStatusShowStopButton() {
    MunkiStatusInfo.shared.status["stop_button_visible"] = true
    postStatusNotification()
}

/// Tells MunkiStatus/MSC to disable the stop button.
func munkiStatusDisableStopButton() {
    MunkiStatusInfo.shared.status["stop_button_enabled"] = false
    postStatusNotification()
}

/// Tells MunkiStatus/MSC to enable the stop button.
func munkiStatusEnableStopButton() {
    MunkiStatusInfo.shared.status["stop_button_enabled"] = true
    postStatusNotification()
}

/// Tells MunkiStatus to bring its window to the front
func munkiStatusActivate() {
    MunkiStatusInfo.shared.status["command"] = "activate"
    postStatusNotification()
    // commands should be cleared after they are sent
    MunkiStatusInfo.shared.status["command"] = ""
}

/// Tells MunkiStatus to quit
func munkiStatusQuit() {
    MunkiStatusInfo.shared.status["command"] = "quit"
    postStatusNotification()
    // commands should be cleared after they are sent
    MunkiStatusInfo.shared.status["command"] = ""
}

/// Tells MunkiStatus to display a restart alert
func munkiStatusRestartAlert() {
    MunkiStatusInfo.shared.status["command"] = "showRestartAlert"
    postStatusNotification()
    // commands should be cleared after they are sent
    MunkiStatusInfo.shared.status["command"] = ""
}
